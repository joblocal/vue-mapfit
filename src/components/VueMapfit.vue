<template>
  <div :id="mapId"></div>
</template>

<script>

export default {
  props: {
    apikey: {
      type: String,
    },
    center: {
      type: [Array, Object],
      required: true,
    },
    theme: {
      type: String,
      default: 'day',
    },
    mapSettings: {
      type: Object,
      default: () => ({}),
    },
  },

  data: () => ({
    mapfitObject: {
      Marker: null,
      MapView: null,
    },
  }),

  computed: {
    hasMapSettings() {
      return Object.keys(this.mapSettings).length > 0;
    },

    mapId() {
      const uid = () => Math.floor((1 + Math.random()) * 0x10000)
        .toString(16)
        .substring(1);

      return ['map', uid(), uid(), uid(), uid()].join('-');
    },
  },

  methods: {
    setMapSettings(map) {
      Object.keys(this.mapSettings).map(setting => map[setting](this.mapSettings[setting]));
    },

    exposeInstances(instances) {
      this.$emit('vueMapfit', instances);
    },

    initMapfit() {
      const { mapfit } = window;
      // draw map
      const map = mapfit.MapView(this.mapId, { theme: this.theme });

      const position = mapfit.LatLng(this.center);
      const marker = mapfit.Marker(position);

      if (this.apikey) mapfit.apikey = this.apikey;

      // set the map center on marker position
      map.setCenter(position);

      // add marker to map
      map.addMarker(marker);

      this.exposeInstances({ map, marker });

      if (this.hasMapSettings) this.setMapSettings(map);
    },

    createScriptTag() {
      const mfScript = document.createElement('script');
      mfScript.setAttribute('type', 'text/javascript');
      mfScript.setAttribute('defer', '');
      mfScript.setAttribute('src', 'https://cdn.mapfit.com/v2-4/assets/js/mapfit.js');

      return mfScript;
    },

    createStyleTag() {
      const mfStyle = document.createElement('link');
      mfStyle.setAttribute('rel', 'stylesheet');
      mfStyle.setAttribute('href', 'https://cdn.mapfit.com/v2-4/assets/css/mapfit.css');

      return mfStyle;
    },
  },

  mounted() {
    if (typeof window === 'undefined') return;

    if (typeof window.mapfit === 'undefined') {
      const style = this.createStyleTag();
      const script = this.createScriptTag();

      script.addEventListener('load', () => this.initMapfit());

      document.head.appendChild(style);
      document.body.appendChild(script);
    } else {
      this.initMapfit();
    }
  },
};
</script>
