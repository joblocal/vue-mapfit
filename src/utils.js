const styleUrl = 'https://cdn.mapfit.com/v2-4/assets/css/mapfit.css';
const scriptUrl = 'https://cdn.mapfit.com/v2-4/assets/js/mapfit.js';
const hasScript = () => Boolean(document.querySelector(`script[src="${scriptUrl}"]`));

const createNode = (tagName, options) => {
  const el = document.createElement(tagName);
  Object.entries(options).forEach(
    ([key, value]) => el.setAttribute(key, value),
  );

  return el;
};

export default function createTags() {
  if (hasScript()) {
    return Promise.resolve(window.mapfit);
  }

  const script = createNode('script', {
    type: 'text/javascript',
    defer: '',
    src: scriptUrl,
  });

  const style = createNode('link', {
    rel: 'stylesheet',
    href: styleUrl,
  });

  document.head.appendChild(style);
  document.body.appendChild(script);

  return new Promise((resolve) => {
    script.addEventListener('load', () => resolve(window.mapfit));
  });
}
