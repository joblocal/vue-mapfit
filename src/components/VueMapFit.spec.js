import td from 'testdouble';
import { shallowMount } from '@vue/test-utils';

import * as utils from '../utils';

// module under test
import VueMapfit from 'src/components/VueMapfit';

const theme = 'day';
const center = {
  lat: 47.850361,
  lng: 12.057812,
};

const propsData = {
  center,
  theme,
};

afterEach(() => {
  td.reset();
});

describe('mounted for mapfit', () => {
  const map = td.object(['addMarker', 'setZoom', 'setCenter']);
  const mapfit = {
    MapView: td.function(),
    LatLng: td.function(),
    Marker: td.function(),
  };

  let wrapper;
  beforeEach(() => {
    const createTags = td.replace(utils, 'createTags');
    td.when(createTags()).thenResolve(mapfit);
  });

  test('to initialize a MapView', async () => {
    const MapView = td.replace(mapfit, 'MapView');
    td.when(MapView(td.matchers.isA(String), { theme })).thenReturn(map);

    wrapper = shallowMount(VueMapfit, { propsData });
    await wrapper.vm.$nextTick();

    td.verify(map.setCenter(td.matchers.anything()));
    td.verify(map.addMarker(td.matchers.anything()));
  });

  test('to set the Mapfit center correctly', async () => {
    td.when(mapfit.MapView(), { ignoreExtraArgs: true }).thenReturn(map);

    const latLng = 'latlng';
    td.when(mapfit.LatLng(center)).thenReturn(latLng);

    wrapper = shallowMount(VueMapfit, { propsData });
    await wrapper.vm.$nextTick();

    td.verify(map.setCenter(latLng));
  });

  test('to set the Mapfit marker correctly', async () => {
    td.when(mapfit.MapView(), { ignoreExtraArgs: true }).thenReturn(map);

    const marker = 'the marker';
    td.when(mapfit.Marker(td.matchers.anything())).thenReturn(marker);

    wrapper = shallowMount(VueMapfit, { propsData });
    await wrapper.vm.$nextTick();

    td.verify(map.addMarker(marker));
  });
});

describe('mounted for mapid', () => {
  const map = td.object(['addMarker', 'setZoom', 'setCenter']);
  const mapfit = {
    MapView: () => map,
    LatLng: td.function(),
    Marker: td.function(),
  };

  let wrapper;
  beforeEach(() => {
    mapfit.apikey = null;

    const createTags = td.replace(utils, 'createTags');
    td.when(createTags()).thenResolve(mapfit);
  });

  test('to set container id signature correctly', async () => {
    wrapper = shallowMount(VueMapfit, { propsData });
    await wrapper.vm.$nextTick();

    const regex = /^map-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}$/i;
    const valid = regex.test(wrapper.attributes().id);
    expect(valid).toBeTruthy();
  });
})

describe('mounted for apiKey', () => {
  const map = td.object(['addMarker', 'setZoom', 'setCenter']);
  const mapfit = {
    MapView: () => map,
    LatLng: td.function(),
    Marker: td.function(),
  };

  let wrapper;
  beforeEach(() => {
    mapfit.apikey = null;

    const createTags = td.replace(utils, 'createTags');
    td.when(createTags()).thenResolve(mapfit);
  });

  test('not to set the mapfit.apiKey when not provided', async () => {
    wrapper = shallowMount(VueMapfit, { propsData });
    await wrapper.vm.$nextTick();

    expect(mapfit.apikey).toBe(null);
  });

  test('to set the mapfit.apiKey when provided', async () => {
    const apikey = '123';
    wrapper = shallowMount(VueMapfit, {
      propsData: { ...propsData, apikey },
    });
    await wrapper.vm.$nextTick();

    expect(mapfit.apikey).toBe(apikey);
  });
});
