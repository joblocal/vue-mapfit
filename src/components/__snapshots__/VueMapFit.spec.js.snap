// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VueMapfit methods generate correct scripts to match snapshot 1`] = `
<div
  id="vue-mapfit"
/>
`;
